flash/core/registry.py

```diff
@@ -0,0 +1,189 @@
+import hashlib
+from collections import defaultdict
+from functools import partial
+from types import FunctionType
+from typing import Any, Callable, Dict, Generator, List, Mapping, Optional, Union
+
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class FlashRegistry(Dict):
+    """
+    This class is used to register function or partial to a registry:
+
+    Example::
+
+        backbones = FlashRegistry("backbones")
+
+        @backbones.register_function()
+        def my_model(nc_input=5, nc_output=6):
+            return nn.Linear(nc_input, nc_output), nc_input, nc_output
+
+        mlp, nc_input, nc_output = backbones.get("my_model")(nc_output=7)
+
+        backbones.register_function(my_model, name="cho")
+        assert backbones.get("cho")
+
+    """
+
+    def __init__(self, registry_name: str, verbose: bool = False):
+        self._registry_name = registry_name
+        self._registered_functions: Mapping[str, Dict[str, Any]] = defaultdict()
+        self._registered_functions_mapping: Dict[str, str] = {}
+        self._verbose = verbose
+
+    def __len__(self):
+        return len(self._registered_functions)
+
+    def __contains__(self, key):
+        return self._registered_functions.get(key, None)
+
+    def __repr__(self):
+        format_str = self.__class__.__name__ + \
+            f'(name={self._registry_name}, ' \
+            f'registered_items={dict(**self._registered_functions)})'
+        return format_str
+
+    @property
+    def name(self) -> str:
+        return self._registry_name
+
+    @property
+    def registered_funcs(self) -> Dict[str, Any]:
+        return self._registered_functions
+
+    def __getitem__(self, key: str) -> Callable:
+        return self.get(key)
```

Comment:
Do we really want 3 different methods to do the same thing? Currently we have `registry.get(backbone)`, `registry(backbone)` and `registry[backbone]` and use all of them through the codebase. IMO we should settle for one option only (discarding the other and make the `get` protected). I'd vote for `registry[backbone]`

Reply:
I would prefer to keep call directly, as registry[backbone, metadata] will looks weird.

Reply:
Keeping call directly.